### PR TITLE
subtyping: apply performance optimization in more places

### PR DESCRIPTION
**Pull request by vtjnash**
~~marked draft, because it'll conflict with #31698, and there will need to be more improvements/corrections to jl_obvious_subtype after that is merged~~

### CODE REVIEW

File: test/core.jl [hunk 1]
```diff
@@ -218,14 +218,14 @@ struct D21923{T,N}; v::D21923{T}; end
...
-    @test Base.isconcretetype(elT2.body.types[1])
-end
+#struct T22624{A,B,C}; v::Vector{T22624{Int64,A}}; end
+#let elT = T22624.body.body.body.types[1].parameters[1]
+#    @test elT == T22624{Int64, T22624.var, C} where C
+#    elT2 = elT.body.types[1].parameters[1]
+#    @test elT2 == T22624{Int64, Int64, C} where C
+#    @test elT2.body.types[1].parameters[1] === elT2
+#    @test Base.isconcretetype(elT2.body.types[1])
+#end
```
Comment (JeffBezanson) [inline test/core.jl]: Why is this disabled?

Comment (vtjnash) [reply test/core.jl]: It’s dependent upon the subtyping fast-path giving a wrong answer (https://github.com/JuliaLang/julia/pull/25796#issuecomment-485797688). I fixed the fast-path and added an assertion in the PR if the fast path diverges from the correct answer, but it causes this test to fail.

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/31807#issuecomment-486770952) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a731d1a74f559bb7cb7df470a52c627918e0b8d4/261e72e_vs_6421def/report.md). cc @ararslan

File: src/subtype.c [hunk 2]
```diff
@@ -1387,6 +1424,28 @@ JL_DLLEXPORT int jl_subtype_env_size(jl_value_t *t)
...
 
+// compute the minimum bound on the number of concrete types that are subtypes of `t`
+// returns 0, 1, or many (2+)
+static int concrete_min(jl_value_t *t)
+{
+    if (jl_is_unionall(t))
+        t = jl_unwrap_unionall(t);
+    if (jl_is_datatype(t)) {
+        if (jl_is_type_type(t))
+            return 0; // Type{T} may have the concrete supertype `typeof(T)`, so don't try to handle them here
```
Comment (JeffBezanson) [inline src/subtype.c]: I thought it was the other way around --- that `typeof(T)` could have `Type{T}` as its supertype?

File: test/core.jl [hunk 3]
```diff
@@ -227,8 +227,8 @@ let ft = Base.datatype_fieldtypes
     @test ft(elT2.body)[1].parameters[1] === elT2
     @test Base.isconcretetype(ft(elT2.body)[1])
 end
-struct S22624{A,B,C} <: Ref{S22624{Int64,A}}; end
-@test @isdefined S22624
+#struct S22624{A,B,C} <: Ref{S22624{Int64,A}}; end
```
Comment (JeffBezanson) [inline test/core.jl]: Looks like we should try to merge #25796 first?

File: src/subtype.c [hunk 4]
```diff
@@ -1637,16 +1771,90 @@ JL_DLLEXPORT int jl_subtype(jl_value_t *x, jl_value_t *y)
...
+        init_stenv(&e, NULL, 0);
+        int subtype = forall_exists_subtype(a, b, &e, 0);
+        assert(subtype_ab == 3 || subtype_ab == subtype || jl_has_free_typevars(a) || jl_has_free_typevars(b));
+#ifndef NDEBUG
+        if (subtype_ab != 0 && subtype_ab != 1) // ensures that running in a debugger doesn't change the result
+#endif
+        subtype_ab = subtype;
+#ifdef NDEBUG
+        if (subtype_ab == 0)
+            return 0;
```
Comment (JeffBezanson) [inline src/subtype.c]: I think this is unreachable?

File: src/subtype.c (hunk 2 above)
Comment (vtjnash) [reply src/subtype.c]:
```
julia> T = Int; Type{T} <: typeof(T)
true
```

Comment (JeffBezanson) [reply src/subtype.c]: Ah, I thought you meant the `.super` field.

File: test/core.jl (hunk 3 above)
Comment (vtjnash) [reply test/core.jl]: I think they end up being somewhat unrelated

**Pull request merged**